secret_store/src/key_server_cluster/client_sessions/generation_session.rs

```diff
@@ -940,406 +940,315 @@ fn check_threshold(threshold: usize, nodes: &BTreeSet<NodeId>) -> Result<(), Err
+		PublicKeyShare, ConfirmInitialization};
+	use key_server_cluster::cluster::tests::{MessagesLoop, make_clusters_and_preserve_sessions};
 	use key_server_cluster::cluster_sessions::ClusterSession;
-	use key_server_cluster::generation_session::{SessionImpl, SessionState, SessionParams};
+	use key_server_cluster::generation_session::{SessionImpl, SessionState};
 	use key_server_cluster::math;
 	use key_server_cluster::math::tests::do_encryption_and_decryption;
 
-	pub struct Node {
-		pub cluster: Arc<DummyCluster>,
-		pub key_storage: Arc<DummyKeyStorage>,
-		pub session: SessionImpl,
-	}
-
-	pub struct MessageLoop {
-		pub session_id: SessionId,
-		pub nodes: BTreeMap<NodeId, Node>,
-		pub queue: VecDeque<(NodeId, NodeId, Message)>,
-	}
-
-	pub fn generate_nodes_ids(n: usize) -> BTreeSet<NodeId> {
-		(0..n).map(|_| math::generate_random_point().unwrap()).collect()
-	}
+	#[derive(Debug)]
+	pub struct MessageLoop(pub MessagesLoop);
 
 	impl MessageLoop {
-		pub fn new(nodes_num: usize) -> Self {
-			Self::with_nodes_ids(generate_nodes_ids(nodes_num))
+		pub fn new(num_nodes: usize) -> Self {
+			MessageLoop(make_clusters_and_preserve_sessions(num_nodes))
 		}
 
-		pub fn with_nodes_ids(nodes_ids: BTreeSet<NodeId>) -> Self {
-			let mut nodes = BTreeMap::new();
-			let session_id = SessionId::default();
-			for node_id in nodes_ids {
-				let cluster = Arc::new(DummyCluster::new(node_id.clone()));
-				let key_storage = Arc::new(DummyKeyStorage::default());
-				let session = SessionImpl::new(SessionParams {
-					id: session_id.clone(),
-					self_node_id: node_id.clone(),
-					key_storage: Some(key_storage.clone()),
-					cluster: cluster.clone(),
-					nonce: Some(0),
-				});
-				nodes.insert(node_id, Node { cluster: cluster, key_storage: key_storage, session: session });
-			}
-
-			let nodes_ids: Vec<_> = nodes.keys().cloned().collect();
-			for node in nodes.values() {
-				for node_id in &nodes_ids {
-					node.cluster.add_node(node_id.clone());
-				}
-			}
-
-			MessageLoop {
-				session_id: session_id,
-				nodes: nodes,
-				queue: VecDeque::new(),
-			}
+		pub fn init(self, threshold: usize) -> Result<Self, Error> {
+			self.0.cluster(0).client().new_generation_session(Default::default(), None, Default::default(), threshold)
+				.map(|_| self)
 		}
 
-		pub fn master(&self) -> &SessionImpl {
-			&self.nodes.values().nth(0).unwrap().session
+		pub fn session_at(&self, idx: usize) -> Arc<SessionImpl> {
+			self.0.sessions(idx).generation_sessions.first().unwrap()
 		}
 
-		pub fn first_slave(&self) -> &SessionImpl {
-			&self.nodes.values().nth(1).unwrap().session
+		pub fn session_of(&self, node: &NodeId) -> Arc<SessionImpl> {
+			self.0.sessions_of(node).generation_sessions.first().unwrap()
 		}
 
-		pub fn second_slave(&self) -> &SessionImpl {
-			&self.nodes.values().nth(2).unwrap().session
+		pub fn take_message_confirm_initialization(&self) -> (NodeId, NodeId, ConfirmInitialization) {
+			match self.0.take_message() {
```

Comment:
`Optional`, but I prefer if/else instead of `pattern match` with only two variants